src/NUnitTestAdapter/NUnitTestDiscoverer.cs

```diff
@@ -39,78 +42,112 @@ public void DiscoverTests(IEnumerable<string> sources, IDiscoveryContext discove
+                        }
                     }
-                    else
+                    catch (BadImageFormatException)
                     {
-                        TestLog.NUnitLoadError(sourceAssembly);
+                        // we skip the native c++ binaries that we don't support.
+                        TestLog.AssemblyNotSupportedWarning(sourceAssembly);
                     }
-                }
-                catch (BadImageFormatException)
-                {
-                    // we skip the native c++ binaries that we don't support.
-                    TestLog.AssemblyNotSupportedWarning(sourceAssembly);
-                }
+                    catch (FileNotFoundException ex)
+                    {
+                        // Probably from the GetExportedTypes in NUnit.core, attempting to find an assembly, not a problem if it is not NUnit here
+                        TestLog.DependentAssemblyNotFoundWarning(ex.FileName, sourceAssembly);
+                    }
+                    catch (FileLoadException ex)
+                    {
+                        // Attempts to load an invalid assembly, or an assembly with missing dependencies
+                        TestLog.LoadingAssemblyFailedWarning(ex.FileName, sourceAssembly);
+                    }
+                    catch (Exception ex)
+                    {
 
-                catch (FileNotFoundException ex)
-                {
-                    // Probably from the GetExportedTypes in NUnit.core, attempting to find an assembly, not a problem if it is not NUnit here
-                    TestLog.DependentAssemblyNotFoundWarning(ex.FileName, sourceAssembly);
-                }
-                catch (FileLoadException ex)
-                {
-                    // Attempts to load an invalid assembly, or an assembly with missing dependencies
-                    TestLog.LoadingAssemblyFailedWarning(ex.FileName, sourceAssembly);
+                        TestLog.SendErrorMessage("Exception thrown discovering tests in " + sourceAssembly, ex);
+                    }
+                    finally
+                    {
+                        if (testConverter != null)
+                            testConverter.Dispose();
+                    }
                 }
-                catch (Exception ex)
-                {
+            }
 
-                    TestLog.SendErrorMessage("Exception thrown discovering tests in " + sourceAssembly, ex);
-                }
-                finally
+            Info("discovering test", "finished");
+        }
+
+        #endregion
+
+        #region Helper Methods
+
+        private bool IsNUnit3TestAssembly(string sourceAssembly)
+        {
+            var assembly = Assembly.ReflectionOnlyLoadFrom(sourceAssembly);
+
+            foreach (var refAssembly in assembly.GetReferencedAssemblies())
+            {
+                if (refAssembly.Name == "nunit.framework")
```

Comment:
I'm always itching to take them out when I find them. :smile: 

It's a matter of personal style, I guess, but I find Linq queries are often not so self-explanatory as their authors find them.